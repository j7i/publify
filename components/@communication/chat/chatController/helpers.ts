import { IMessage } from '../types'

export const sortMessageByTimeStamp = (messages: IMessage[]): IMessage[] =>
  messages.sort((a: IMessage, b: IMessage) => {
    let timeA = `${a.date.seconds}${a.date.nanoseconds}`
    let timeB = `${b.date.seconds}${b.date.nanoseconds}`
    return parseInt(timeA, 10) - parseInt(timeB, 10)
  })
