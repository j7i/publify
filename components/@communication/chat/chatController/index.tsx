import { firestore } from '@config/firebase'
import { FirebaseCollection } from '@config/firebase/types.d'
import { PureComponent, ReactNode } from 'react'
import { IChatControllerProps, IChatControllerState, IChatRenderProps, IMessage } from './types'

export default class ChatController extends PureComponent<IChatControllerProps, IChatControllerState> {
  public state: IChatControllerState = {
    message: '',
    messages: [],
    loading: true
  }

  public componentDidMount(): void {
    const { seekingId, seekingOwnerId, loggedInUser } = this.props
    const firestoreChat = firestore.collection('chats').doc(seekingId + loggedInUser)

    // seekingOwnerId is possibly undefined
    firestoreChat
      .get()
      .then((docSnapshot: firebase.firestore.DocumentSnapshot) => {
        if (docSnapshot.exists) {
          this.startChatConversation(seekingId, loggedInUser)
        } else {
          firestoreChat
            .set({
              members: [seekingOwnerId, loggedInUser]
            })
            .then(() => {
              this.startChatConversation(seekingId, loggedInUser)
            })
        }
      })
      .catch((error: Error) => {
        // tslint:disable-next-line:no-console
        console.error('Error initializing chat: ', error)
      })
  }

  public startChatConversation(seekingId: string, loggedInUser: string): void {
    firestore
      .collection('chats')
      .doc(seekingId + loggedInUser)
      .collection('messages')
      .onSnapshot(
        (querySnapshot: firebase.firestore.QuerySnapshot) => {
          let messages: firebase.firestore.DocumentData = []
          querySnapshot.forEach((doc: firebase.firestore.QueryDocumentSnapshot) => {
            messages.push(doc.data())
          })
          messages.sort((a: IMessage, b: IMessage) => {
            return a.date.seconds - b.date.seconds
          })
          this.setState({
            messages,
            loading: false
          })
        },
        (error: Error) => {
          // tslint:disable-next-line:no-console
          console.error(error)
        }
      )
  }

  public componentWillUnmount(): void {
    const { seekingId } = this.props

    firestore
      .collection('chats')
      .doc(seekingId)
      .collection('messages')
      // tslint:disable-next-line:no-empty
      .onSnapshot(() => {})
  }

  public sendMessage = (event: React.FormEvent<HTMLFormElement>): void => {
    event.preventDefault()

    const { seekingId, loggedInUser } = this.props
    const message = {
      content: this.state.message,
      date: new Date(),
      uid: loggedInUser
    }

    this.setState({
      message: ''
    })

    firestore
      .collection(FirebaseCollection.CHATS)
      .doc(seekingId + loggedInUser)
      .collection('messages')
      .add({ ...message })
      .then(() => {
        // tslint:disable-next-line:no-console
        console.log('Message sent')
      })
      .catch((error: Error) => {
        // tslint:disable-next-line:no-console
        console.error('Error adding message: ', error)
      })
  }

  public handleChange = (event: React.ChangeEvent<HTMLInputElement>): void => {
    const { value } = event.target

    event.persist()
    this.setState({
      message: value
    })
  }

  public render(): ReactNode {
    const { children, loggedInUser } = this.props

    const chatRenderProps: IChatRenderProps = {
      ...this.state,
      loggedInUser,
      handleChange: this.handleChange,
      sendMessage: this.sendMessage
    }

    return children(chatRenderProps)
  }
}
