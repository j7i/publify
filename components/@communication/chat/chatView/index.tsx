import { PureComponent, ReactNode } from 'react'
import ChatConversation from './conversation'
import ChatHeader from './header'
import styles from './styles.css'
import ChatTrigger from './trigger'
import { IChatViewProps } from './types'

export default class ChatView extends PureComponent<IChatViewProps> {
  public render(): ReactNode {
    const { messages, handleChange, sendMessage } = this.props.chatRenderProps

    let fetchedMessages
    if (messages) {
      fetchedMessages = messages
    }

    return (
      <section className={styles.chat}>
        <div className={styles.chatInner}>
          <ChatHeader />
          <ChatConversation fetchedMessages={fetchedMessages} />
          <ChatTrigger handleChange={handleChange} sendMessage={sendMessage} />
        </div>
      </section>
    )
  }
}
