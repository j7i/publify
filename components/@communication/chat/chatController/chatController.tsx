import { FirebaseCollection, firestore } from '@config'
import { PureComponent, ReactNode } from 'react'
import { IChatControllerProps, IChatControllerState, IChatViewRenderProps, IMemberInfo, IMessage } from '../types'
import { sortMessageByTimeStamp } from './helpers'

export class ChatController extends PureComponent<IChatControllerProps, IChatControllerState> {
  public state: IChatControllerState = {
    message: '',
    messages: [],
    loading: true
  }

  public componentDidMount(): void {
    const { advertId, advertTitle, advertOwnerId, advertOwnerName, advertOwnerImageURL, loggedInUserId, loggedInUserName, loggedInUserImageURL } = this.props

    this.firestoreChat()
      .get()
      .then((docSnapshot: firebase.firestore.DocumentSnapshot) => {
        if (docSnapshot.exists) {
          this.startChatConversation()
        } else if (advertOwnerId) {
          const memberInfos: IMemberInfo = {
            [advertOwnerId]: { name: advertOwnerName!, userImageURL: null },
            [loggedInUserId]: { name: loggedInUserName, userImageURL: null }
          }

          if (advertOwnerImageURL) {
            memberInfos[advertOwnerId].userImageURL = advertOwnerImageURL
          }

          if (loggedInUserImageURL) {
            memberInfos[loggedInUserId].userImageURL = loggedInUserImageURL
          }

          this.firestoreChat()
            .set({
              memberInfos,
              members: [advertOwnerId, loggedInUserId],
              advertId,
              advertTitle
            })
            .then(() => {
              this.startChatConversation()
            })
        }
      })
      .catch((error: Error) => {
        // tslint:disable-next-line:no-console
        console.error('Error initializing chat: ', error)
      })
  }

  public componentWillUnmount(): void {
    this.firestoreChat()
      .collection(FirebaseCollection.CHAT_MESSAGES)
      // tslint:disable-next-line:no-empty
      .onSnapshot(() => {})
  }

  public handleChange = (event: React.ChangeEvent<HTMLInputElement>): void => {
    event.persist()

    const { value } = event.target

    this.setState({
      message: value
    })
  }

  public sendMessage = (event: React.FormEvent<HTMLFormElement>): void => {
    event.preventDefault()

    const { loggedInUserId } = this.props
    const message = {
      content: this.state.message,
      date: new Date(),
      uid: loggedInUserId
    }

    this.setState({
      message: ''
    })

    this.firestoreChat()
      .collection(FirebaseCollection.CHAT_MESSAGES)
      .add({ ...message })
      .catch((error: Error) => {
        // tslint:disable-next-line:no-console
        console.error('Error adding message: ', error)
      })
  }

  public render(): ReactNode {
    const { children, loggedInUserId } = this.props

    const chatViewRenderProps: IChatViewRenderProps = {
      ...this.state,
      loggedInUserId,
      handleChange: this.handleChange,
      sendMessage: this.sendMessage
    }

    return children(chatViewRenderProps)
  }

  private firestoreChat = (): firebase.firestore.DocumentReference => {
    const { advertId, loggedInUserId, chatId } = this.props
    const chat = advertId && loggedInUserId ? advertId + loggedInUserId : chatId

    return firestore.collection(FirebaseCollection.CHATS).doc(chat)
  }

  private startChatConversation = (): void => {
    this.firestoreChat()
      .collection(FirebaseCollection.CHAT_MESSAGES)
      .onSnapshot(
        (querySnapshot: firebase.firestore.QuerySnapshot) => {
          let messages: firebase.firestore.DocumentData = []

          querySnapshot.forEach((doc: firebase.firestore.QueryDocumentSnapshot) => {
            messages.push(doc.data())
          })

          messages = sortMessageByTimeStamp(messages as IMessage[])

          this.setState({
            messages,
            loading: false
          })
        },
        (error: Error) => {
          // tslint:disable-next-line:no-console
          console.error(error)
        }
      )
  }
}
