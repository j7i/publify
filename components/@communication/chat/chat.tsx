import { UserSpecificContent } from '@auth'
import { IChatRenderProps } from '@communication/chat/types'
import { IUserInfo } from '@user'
import { PureComponent, ReactNode } from 'react'
import { ChatController } from './chatController/chatController'
import { ChatView } from './chatView/chatView'
import { IChatProps } from './types'

export class Chat extends PureComponent<IChatProps> {
  public render(): JSX.Element {
    const { advertId, advertTitle, advertOwnerId, advertOwnerName, chatId, displayMode } = this.props

    return (
      <UserSpecificContent>
        {(user: firebase.User, userInfo: IUserInfo): ReactNode => {
          return user ? (
            <ChatController
              advertId={advertId}
              advertTitle={advertTitle}
              advertOwnerId={advertOwnerId}
              advertOwnerName={advertOwnerName}
              loggedInUserId={user.uid}
              loggedInUserName={userInfo.fullName}
              chatId={chatId}
            >
              {(chatRenderProps: IChatRenderProps): ReactNode => {
                return <ChatView chatRenderProps={chatRenderProps} displayMode={displayMode} />
              }}
            </ChatController>
          ) : (
            <>TODO</>
          )
        }}
      </UserSpecificContent>
    )
  }
}
