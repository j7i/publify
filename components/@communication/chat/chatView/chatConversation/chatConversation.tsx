import { IChatConversationProps, IMessage } from '@communication/chat/types'
import CircularProgress from '@material-ui/core/CircularProgress'
import classNames from 'classnames'
import { createRef, PureComponent, ReactNode, RefObject } from 'react'
import Moment from 'react-moment'
import styles from './chatConversationStyles.css'

export class ChatConversation extends PureComponent<IChatConversationProps> {
  private conversationEnd: RefObject<HTMLDivElement> = createRef<HTMLDivElement>()

  public componentDidUpdate(): void {
    const conversationEnd = this.conversationEnd.current!
    conversationEnd.scrollIntoView({ block: 'end', behavior: 'smooth' })
  }

  public render(): ReactNode {
    const { loading, fetchedMessages, loggedInUserId } = this.props

    return (
      <div
        className={classNames(styles.conversation, styles.chat, {
          [styles.isLoading]: loading
        })}
      >
        {loading ? (
          <CircularProgress className={styles.loadingSpinner} />
        ) : (
          <div className={styles.scrollable}>
            {fetchedMessages &&
              fetchedMessages.map((message: IMessage, index: number) => (
                <div
                  key={index}
                  className={classNames(styles.messages, {
                    [styles.others]: message.uid !== loggedInUserId,
                    [styles.self]: message.uid === loggedInUserId
                  })}
                >
                  <div className={styles.text}>
                    {message.content}
                    <span className={styles.time}>{message.date && <Moment format="HH:mm" unix date={new Date(message.date.seconds)} />}</span>
                  </div>
                </div>
              ))}
            <div ref={this.conversationEnd} />
          </div>
        )}
      </div>
    )
  }
}
