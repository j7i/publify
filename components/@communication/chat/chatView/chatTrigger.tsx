import { IChatTriggerProps } from '@communication/chat/types'
import IconButton from '@material-ui/core/IconButton'
import Send from '@material-ui/icons/Send'
import { PureComponent, ReactNode } from 'react'
import styles from './chatViewStyles.css'

export class ChatTrigger extends PureComponent<IChatTriggerProps> {
  public render(): ReactNode {
    const { handleChange, sendMessage, message } = this.props
    return (
      <form className={styles.chatTrigger} onSubmit={sendMessage}>
        <input type="text" placeholder="Write a message" value={message} onChange={handleChange} />
        <IconButton color="primary" type="submit" disabled={message.length === 0}>
          <Send />
        </IconButton>
      </form>
    )
  }
}
