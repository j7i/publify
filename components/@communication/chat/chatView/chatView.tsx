import { IChatViewProps } from '@communication/chat/types'
import { PureComponent, ReactNode } from 'react'
import { ChatConversation } from './chatConversation/chatConversation'
// import { ChatHeader } from './chatHeader'
import { ChatTrigger } from './chatTrigger'
import styles from './chatViewStyles.css'

export class ChatView extends PureComponent<IChatViewProps> {
  public render(): ReactNode {
    const { message, messages, loading, handleChange, sendMessage, loggedInUserId } = this.props.chatViewRenderProps

    let fetchedMessages
    if (messages) {
      fetchedMessages = messages
    }

    return (
      <section className={styles.chat}>
        <div className={styles.chatInner}>
          <ChatConversation loading={loading} fetchedMessages={fetchedMessages} loggedInUserId={loggedInUserId} />
          <ChatTrigger message={message} handleChange={handleChange} sendMessage={sendMessage} />
        </div>
      </section>
    )
  }
}
