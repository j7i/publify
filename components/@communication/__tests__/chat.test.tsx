/* eslint-env jest */
import { cleanup, fireEvent, render } from 'react-testing-library'
import { basicFetchedMessagesMock, sortedFetchedMessagesMock, unsortedFetchedMessagesMock } from '../__mocks__/chat.mock'
import { sortMessageByTimeStamp } from '../chat/chatController/helpers'
import { ChatConversation } from '../chat/chatView/chatConversation/chatConversation'
import { ChatTrigger } from '../chat/chatView/chatTrigger'

describe('@communication/chat:', () => {
  describe('ChatController', () => {
    it('should provide sorted messages ', () => {
      // arrange
      const fetchedMessages = unsortedFetchedMessagesMock
      const expectedMessages = sortedFetchedMessagesMock

      // act
      const sortedMessages = sortMessageByTimeStamp(fetchedMessages)

      // assert
      expect(sortedMessages).toStrictEqual(expectedMessages)
    })
  })

  describe('ChatConversation', () => {
    afterEach(cleanup)

    it('should render correctly', () => {
      const { container } = render(<ChatConversation loading={false} loggedInUserId={'1a2b'} fetchedMessages={basicFetchedMessagesMock} />)
      const totalMessages = container.querySelectorAll('.messages')
      const ownMessages = container.querySelectorAll('.self')

      expect(container).toMatchSnapshot('Representing a basic chat conversation with two participants.')
      expect(totalMessages.length).toBe(3)
      expect(ownMessages.length).toBe(2)
    })

    it('should render a loading animation while loading', () => {
      const { container } = render(<ChatConversation loading={true} loggedInUserId={'1a2b'} fetchedMessages={basicFetchedMessagesMock} />)

      expect(container).toMatchSnapshot('Representing a loading chat conversation.')
    })
  })

  describe('ChatTrigger', () => {
    afterEach(cleanup)

    it('should handle submit correctly', () => {
      const handleChangeMock = jest.fn()
      const sendMessageMock = jest.fn()
      const messageMock = `I'm a message mock as string`

      const { container } = render(<ChatTrigger handleChange={handleChangeMock} sendMessage={sendMessageMock} message={messageMock} />)
      const form = container.querySelector('form') as HTMLElement
      fireEvent.submit(form)

      expect(sendMessageMock).toHaveBeenCalledTimes(1)
    })

    it('should handle message input change correctly', () => {
      const handleChangeMock = jest.fn()
      const sendMessageMock = jest.fn()
      const messageMock = 'I will be the change trigger'

      const { container } = render(<ChatTrigger handleChange={handleChangeMock} sendMessage={sendMessageMock} message={''} />)
      const input = container.querySelector('input') as HTMLInputElement
      fireEvent.change(input, { target: { value: messageMock } })

      expect(handleChangeMock).toHaveBeenCalledTimes(1)
    })
  })
})
