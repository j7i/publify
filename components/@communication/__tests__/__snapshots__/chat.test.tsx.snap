// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@communication/chat: ChatConversation should render a loading animation while loading: Representing a loading chat conversation. 1`] = `
<div>
  <div
    class="conversation chat isLoading"
  >
    <div
      class="MuiCircularProgress-root-1 MuiCircularProgress-colorPrimary-4 MuiCircularProgress-indeterminate-3 loadingSpinner"
      role="progressbar"
      style="width: 40px; height: 40px;"
    >
      <svg
        class="MuiCircularProgress-svg-6"
        viewBox="22 22 44 44"
      >
        <circle
          class="MuiCircularProgress-circle-7 MuiCircularProgress-circleIndeterminate-9"
          cx="44"
          cy="44"
          fill="none"
          r="20.2"
          stroke-width="3.6"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`@communication/chat: ChatConversation should render correctly: Representing a basic chat conversation with two participants. 1`] = `
<div>
  <div
    class="conversation chat"
  >
    <div
      class="scrollable"
    >
      <div
        class="messages self"
      >
        <div
          class="text"
        >
          I wrote this message, so it is mine.
          <span
            class="time"
          >
            <time
              datetime="1548022013000"
            >
              23:06
            </time>
          </span>
        </div>
      </div>
      <div
        class="messages others"
      >
        <div
          class="text"
        >
          So then I can have mine too..
          <span
            class="time"
          >
            <time
              datetime="1548022016000"
            >
              23:06
            </time>
          </span>
        </div>
      </div>
      <div
        class="messages self"
      >
        <div
          class="text"
        >
          Yeah sure,s we are kind of the same :)
          <span
            class="time"
          >
            <time
              datetime="1548022020000"
            >
              23:07
            </time>
          </span>
        </div>
      </div>
      <div />
    </div>
  </div>
</div>
`;
